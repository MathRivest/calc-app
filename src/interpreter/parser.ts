import {
  Token,
  SyntaxKind,
  NumberToken,
  BinaryLiteralToken,
  UnitToken,
} from './tokenizer';
import { Representation } from './interpreter';

export type Node = Expression;

type MathOperator = '+' | '-' | '*' | '/' | '^';

export enum NodeKind {
  NumberLiteral,
  BinaryExpression,
  UnaryPlus,
  UnaryMinus,
  ConvertToBinary,
  ConvertToDecimal,
  ConvertToUnit,
}

export type Expression =
  | NumberLiteral
  | UnaryMinus
  | UnaryPlus
  | BinaryExpression
  | ConvertToBinary
  | ConvertToDecimal
  | ConvertToUnit;

export type NumberLiteral = {
  kind: NodeKind.NumberLiteral;
  value: number;
  representation: Representation;
};

export type ConvertToUnit = {
  kind: NodeKind.ConvertToUnit;
  unit: string;
  expression: Expression;
};

export type UnaryPlus = {
  kind: NodeKind.UnaryPlus;
  expression: Expression;
};

export type UnaryMinus = {
  kind: NodeKind.UnaryMinus;
  expression: Expression;
};

export type BinaryExpression = {
  kind: NodeKind.BinaryExpression;
  left: Expression;
  right: Expression;
  operator: MathOperator;
};

export type ConvertToBinary = {
  kind: NodeKind.ConvertToBinary;
  expression: Expression;
};

export type ConvertToDecimal = {
  kind: NodeKind.ConvertToDecimal;
  expression: Expression;
};

export default function parser(tokens: Token[]): Node {
  let i = 0;

  function currentToken(): Token {
    return tokens[i];
  }

  function testAndConsume(kind: SyntaxKind): boolean {
    const token = currentToken();
    if (token.kind === kind) {
      i++;
      return true;
    }
    return false;
  }

  function consumeToken(kind: SyntaxKind): Token {
    const token = tokens[i];
    if (token.kind !== kind) {
      throw new Error(`Unexpected token: ${token}`);
    }
    i++;
    return token;
  }

  function makeNumberLiteral(token: NumberToken): NumberLiteral {
    return {
      kind: NodeKind.NumberLiteral,
      value: parseFloat(token.value),
      representation: Representation.Unknown,
    };
  }

  function makeNumberLiteralFromBinary(
    token: BinaryLiteralToken
  ): NumberLiteral {
    return {
      kind: NodeKind.NumberLiteral,
      value: parseInt(token.value, 2),
      representation: Representation.Binary,
    };
  }

  function makeBinaryExpression(
    left: Expression,
    operator: MathOperator,
    right: Expression
  ): BinaryExpression {
    return {
      kind: NodeKind.BinaryExpression,
      left: left,
      operator: operator,
      right: right,
    };
  }

  function factor(): Expression {
    let token = currentToken();
    if (testAndConsume(SyntaxKind.PlusToken)) {
      return { kind: NodeKind.UnaryPlus, expression: factor() };
    } else if (testAndConsume(SyntaxKind.MinusToken)) {
      return { kind: NodeKind.UnaryMinus, expression: factor() };
    } else if (token.kind === SyntaxKind.BinaryLiteral) {
      return makeNumberLiteralFromBinary(consumeToken(
        SyntaxKind.BinaryLiteral
      ) as BinaryLiteralToken);
    } else if (token.kind === SyntaxKind.NumberLiteral) {
      return makeNumberLiteral(consumeToken(
        SyntaxKind.NumberLiteral
      ) as NumberToken);
    } else if (testAndConsume(SyntaxKind.LPrecedence)) {
      let node = expr();
      consumeToken(SyntaxKind.RPrecedence);
      return node;
    }
    throw Error(`Unexpected token : ${token}`);
  }

  function unit(): Expression {
    let node: Expression = factor();
    if (currentToken().kind === SyntaxKind.Unit) {
      const unitToken = consumeToken(SyntaxKind.Unit) as UnitToken;
      node = {
        kind: NodeKind.ConvertToUnit,
        expression: node,
        unit: unitToken.value,
      } as ConvertToUnit;
    }
    return node;
  }

  function conversion(): Expression {
    let node: Expression = unit();

    while (true) {
      if (testAndConsume(SyntaxKind.In)) {
        if (testAndConsume(SyntaxKind.Binary)) {
          node = {
            kind: NodeKind.ConvertToBinary,
            expression: node,
          };
        } else if (testAndConsume(SyntaxKind.Decimal)) {
          node = {
            kind: NodeKind.ConvertToDecimal,
            expression: node,
          };
        } else if (currentToken().kind === SyntaxKind.Unit) {
          const unitToken = consumeToken(SyntaxKind.Unit) as UnitToken;
          node = {
            kind: NodeKind.ConvertToUnit,
            expression: node,
            unit: unitToken.value,
          };
        } else {
          throw new Error(`Can't extract unit from token ${currentToken()}`);
        }
      } else {
        break;
      }
    }

    return node;
  }

  function exponent(): Expression {
    let node: Expression = conversion();

    while (testAndConsume(SyntaxKind.CaretToken)) {
      node = makeBinaryExpression(node, '^', conversion());
    }

    return node;
  }

  function term(): Expression {
    let node: Expression = exponent();

    while (true) {
      if (testAndConsume(SyntaxKind.AsteriskToken)) {
        node = makeBinaryExpression(node, '*', exponent());
      } else if (testAndConsume(SyntaxKind.SlashToken)) {
        node = makeBinaryExpression(node, '/', exponent());
      } else if (testAndConsume(SyntaxKind.LPrecedence)) {
        let ex = expr();
        consumeToken(SyntaxKind.RPrecedence);
        node = makeBinaryExpression(node, '*', ex);
      } else {
        break;
      }
    }

    return node;
  }

  function expr(): Expression {
    let node: Expression = term();

    while (true) {
      if (testAndConsume(SyntaxKind.PlusToken)) {
        node = makeBinaryExpression(node, '+', term());
      } else if (testAndConsume(SyntaxKind.MinusToken)) {
        node = makeBinaryExpression(node, '-', term());
      } else {
        break;
      }
    }

    return node;
  }

  return expr();
}
