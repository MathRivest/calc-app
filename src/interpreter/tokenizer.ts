export interface Token {
  kind: SyntaxKind;
}

export enum SyntaxKind {
  Number,
  EOF,
  LPrecedence,
  RPrecedence,
  PlusToken,
  MinusToken,
  AsteriskToken,
  SlashToken,
  CaretToken,
}

export interface NumberToken extends Token {
  kind: SyntaxKind.Number;
  value: string;
}

const simpleCharToSyntaxKindMap: Map<string, SyntaxKind> = new Map([
  ['+', SyntaxKind.PlusToken],
  ['-', SyntaxKind.MinusToken],
  ['*', SyntaxKind.AsteriskToken],
  ['/', SyntaxKind.SlashToken],
  ['(', SyntaxKind.LPrecedence],
  [')', SyntaxKind.RPrecedence],
  ['^', SyntaxKind.CaretToken],
]);

function isDigit(char: string) {
  const NUMBERS = /[0-9]/;
  return NUMBERS.test(char);
}

export default function tokenizer(input: string): Token[] {
  let current = 0;

  const tokens: Token[] = [];

  while (current < input.length) {
    let char = input[current];

    let syntaxKind = simpleCharToSyntaxKindMap.get(char);
    if (syntaxKind !== undefined) {
      tokens.push({
        kind: syntaxKind,
      });
      current++;
    } else if (isDigit(char)) {
      let value = '';
      while (isDigit(char)) {
        value += char;
        char = input[++current];
      }
      tokens.push({ kind: SyntaxKind.Number, value } as NumberToken);
    } else if (char == ' ') {
      current++;
    } else {
      throw new Error('Unkown character: ' + char);
    }
  }

  tokens.push({ kind: SyntaxKind.EOF });
  return tokens;
}
