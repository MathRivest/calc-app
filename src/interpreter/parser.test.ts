import tokenizer from './tokenizer';
import parser from './parser';

test('should match addition snapshot', () => {
  const ast = parser(tokenizer('1+2'));
  expect(ast).toMatchSnapshot();
});

test('should match substraction snapshot', () => {
  const ast = parser(tokenizer('1-2'));
  expect(ast).toMatchSnapshot();
});

test('should match multiplication snapshot', () => {
  const ast = parser(tokenizer('1*2'));
  expect(ast).toMatchSnapshot();
});

test('should match division snapshot', () => {
  const ast = parser(tokenizer('1/2'));
  expect(ast).toMatchSnapshot();
});

test('should match chain addition snapshot', () => {
  const ast = parser(tokenizer('1+2-2'));
  expect(ast).toMatchSnapshot();
});

test('should match multiplication and addition snapshot', () => {
  const ast = parser(tokenizer('1+2*2'));
  expect(ast).toMatchSnapshot();
});

test('should match precedence snapshot', () => {
  const ast = parser(tokenizer('(1+2)*2'));
  expect(ast).toMatchSnapshot();
});
