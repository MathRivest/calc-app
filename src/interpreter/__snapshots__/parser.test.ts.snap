// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match addition snapshot 1`] = `
Object {
  "kind": 6,
  "left": Object {
    "kind": 0,
    "representation": 0,
    "value": 1,
  },
  "right": Object {
    "kind": 0,
    "representation": 0,
    "value": 2,
  },
}
`;

exports[`should match chain addition snapshot 1`] = `
Object {
  "kind": 7,
  "left": Object {
    "kind": 6,
    "left": Object {
      "kind": 0,
      "representation": 0,
      "value": 1,
    },
    "right": Object {
      "kind": 0,
      "representation": 0,
      "value": 2,
    },
  },
  "right": Object {
    "kind": 0,
    "representation": 0,
    "value": 2,
  },
}
`;

exports[`should match division snapshot 1`] = `
Object {
  "kind": 9,
  "left": Object {
    "kind": 0,
    "representation": 0,
    "value": 1,
  },
  "right": Object {
    "kind": 0,
    "representation": 0,
    "value": 2,
  },
}
`;

exports[`should match multiplication and addition snapshot 1`] = `
Object {
  "kind": 6,
  "left": Object {
    "kind": 0,
    "representation": 0,
    "value": 1,
  },
  "right": Object {
    "kind": 8,
    "left": Object {
      "kind": 0,
      "representation": 0,
      "value": 2,
    },
    "right": Object {
      "kind": 0,
      "representation": 0,
      "value": 2,
    },
  },
}
`;

exports[`should match multiplication snapshot 1`] = `
Object {
  "kind": 8,
  "left": Object {
    "kind": 0,
    "representation": 0,
    "value": 1,
  },
  "right": Object {
    "kind": 0,
    "representation": 0,
    "value": 2,
  },
}
`;

exports[`should match precedence snapshot 1`] = `
Object {
  "kind": 8,
  "left": Object {
    "kind": 6,
    "left": Object {
      "kind": 0,
      "representation": 0,
      "value": 1,
    },
    "right": Object {
      "kind": 0,
      "representation": 0,
      "value": 2,
    },
  },
  "right": Object {
    "kind": 0,
    "representation": 0,
    "value": 2,
  },
}
`;

exports[`should match substraction snapshot 1`] = `
Object {
  "kind": 7,
  "left": Object {
    "kind": 0,
    "representation": 0,
    "value": 1,
  },
  "right": Object {
    "kind": 0,
    "representation": 0,
    "value": 2,
  },
}
`;
